domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt

```diff
@@ -0,0 +1,131 @@
+package ua.marchenko.artauction.artwork.repository.impl
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.lettuce.core.RedisException
+import java.net.SocketException
+import java.time.Duration
+import org.slf4j.LoggerFactory
+import org.springframework.dao.QueryTimeoutException
+import org.springframework.data.redis.RedisConnectionFailureException
+import org.springframework.data.redis.core.ReactiveRedisTemplate
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+import reactor.kotlin.core.publisher.toMono
+import ua.marchenko.artauction.artwork.model.MongoArtwork
+import ua.marchenko.artauction.artwork.model.projection.ArtworkFull
+import ua.marchenko.artauction.artwork.repository.ArtworkRepository
+import ua.marchenko.core.artwork.enums.ArtworkStatus
+import ua.marchenko.core.artwork.exception.ArtworkNotFoundException
+
+@Repository
+internal class RedisArtworkRepository(
+    private val mongoArtworkRepository: MongoArtworkRepository,
+    private val reactiveRedisTemplate: ReactiveRedisTemplate<String, ByteArray>,
+    private val objectMapper: ObjectMapper,
+) : ArtworkRepository by mongoArtworkRepository {
+
+    override fun save(artwork: MongoArtwork): Mono<MongoArtwork> =
+        mongoArtworkRepository.save(artwork)
+            .flatMap { savedArtwork ->
+                val id = requireNotNull(savedArtwork.id) { "artwork id cannot be null" }.toHexString()
+                reactiveRedisTemplate.delete(createFullKeyById(id))
+                    .then(saveArtworkToRedis(savedArtwork))
+                    .thenReturn(savedArtwork)
+                    .onErrorResume(::isErrorFromRedis) { savedArtwork.toMono() }
```

Comment:
What if:

1. we saved a data to mongo and redis
2. we requested the data and received the cached value from redis
3. we updated the data in mongo and failed to update redisc
4. we requested the data and received the outdated cached value from redis

Reply:
User would receive the outdated values while the cache in redis exist (TTL 10 minutes in my case). Should I try to retry updating data in redis?

Reply:
I have added retries (5 with delay of 2 sec) to update keys in redis after saving/updating operations, but if redis is unavailable more than retries time, the user will recieve the outdated cached (while there are exist due to TTL)

Reply:
changed to backoff with 5 attempts and min delay of 200 millis